### ISSUE TITLE
Stream Freezes with high VRAM utilization (and does not recover)

### ISSUE THREAD

**Issue by Nonary** [LizardByte/Sunshine#1031]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your issue described in the documentation?

- [X] I have read the documentation

### Is your issue present in the nightly release?

- [X] This issue is present in the nightly release

### Describe the Bug

If the host happens to use around 90% VRAM utilization, the stream will run into performance problems. At first it will drop framerate down, but if VRAM is exceeded it will break the stream entirely by causing it to freeze. Users will have to restart sunshine entirely in order to recover the stream, even if host has VRAM available.

This impacts games that use DLSS3 more than any other games because people bought the 12GB VRAM cards and try to push the limits of their card by turning everything up (which DLSS3 can do at reasonable framerates) so it throws them off guard.

Most common offenders are Diablo IV, Hogwarts Legacy and Spiderman. 
Basically any game that uses a lot of VRAM.

### Expected Behavior

If VRAM is exceeded, Sunshine should be able to recover without requiring users to restart it.

### Host Operating System

Windows

### Operating System Version

11

### Architecture

64 bit

### Sunshine commit or version

0.18.4

### Package

Windows - installer

### GPU Type

Nvidia

### GPU Model

N/A

### GPU Driver/Mesa Version

N/A

### Config

```Shell
N/A
```

### Relevant log output

```shell
N/A
```

**Comment by laGrossePistache** [comment]
I've the same problem. I really hope it's gonna be fixe soon !

**Comment by cgutman** [comment]
Likely fixed by #1098. Please retest the latest nightly.

**Comment by Nonary** [comment]
> Likely fixed by #1098. Please retest the latest nightly.

Sadly, even with the changes it still freezes the moment you exceed your VRAM.

Here's a test script using PyCuda (and NVIDIA CUDA API)

```
import pycuda.autoinit
import pycuda.driver as cuda_driver
import numpy as np

def allocate_and_fill_vram(vram_to_allocate_gb):
    vram_to_allocate_bytes = vram_to_allocate_gb * 1024**3
    num_elements = vram_to_allocate_bytes // np.dtype(np.float32).itemsize

    # Create a list to store allocated GPU memory objects
    allocated_gpu_mem = []

    try:
        # Calculate the chunk size and the number of chunks to transfer
        chunk_size_bytes = (1024 * 1)**2  # 1 MiB
        chunk_size_elements = chunk_size_bytes // np.dtype(np.float32).itemsize
        num_chunks = vram_to_allocate_bytes // chunk_size_bytes

        # Transfer data in chunks to avoid exceeding system memory
        for i in range(num_chunks):
            # Allocate pagelocked host memory
            data = cuda_driver.pagelocked_empty(chunk_size_elements, dtype=np.float32)

            # Fill the pagelocked host memory with random data
            np.copyto(data, np.random.randn(chunk_size_elements).astype(np.float32))

            # Allocate GPU memory
            gpu_mem = cuda_driver.mem_alloc(chunk_size_elements * np.dtype(np.float32).itemsize)
            allocated_gpu_mem.append(gpu_mem)

            # Transfer data from pagelocked host memory to GPU memory
            cuda_driver.memcpy_htod(gpu_mem, data)

        print(f"Filled {vram_to_allocate_gb} GB of VRAM with random data")

    except cuda_driver.MemoryError as e:
        print("Memory allocation or data transfer failed. VRAM might be exceeded.")

    return allocated_gpu_mem

if __name__ == "__main__":
    allocated_gpu_mem = allocate_and_fill_vram(21)

    # Keep the script running to prevent GPU memory deallocation
    print("Press Ctrl+C to exit")
    try:
        while True:
            pass
    except KeyboardInterrupt:
        print("Exiting")
```

Using that at desktop and browsing the internet, the moment my VRAM exceeded 24GB it froze immediately and did not recover even after the VRAM was released.

**Comment by kleprevost** [comment]
+1 - this issue is reproducible on my setup using the latest Sunshine nightly build (0.19.1.010440dbd3e0e89e2dd891f0179d126745fa3460). 

I'm running a 3070 with 8GB VRAM and have been experiencing unrecoverable freezes (stuttering when over 90%, then picture freezes while audio continues to play) while playing Resident Evil 4 Remake with high settings. The issue only occurs when VRAM usage exceeds 90%. I can reproduce the issue using the script above. I've also ruled out other variables - network stats look fine in the moonlight console, CPU% and GPU% are under 70%.

**Comment by ns6089** [comment]
At this point I'm starting to think all these new dx12 games with unevictable vram allocations are capable of breaking windows side of desktop duplication api, so it starts returning timeouts instead of surfaces. Needs testing, and if confirmed we can simply reinit capture on prolonged timeouts.

**Comment by svenlange2** [comment]
Happens also in MSFS 2020 in DX11 mode. As the VRAM gets high, stutters and then freeze. Sunshine actually recovers after some 30s but for that it does a restart i guess (in service mode) but when I log back in it crashes again. Only soution is to reduce VRAM usage by changing settings in game.

**Comment by ns6089** [comment]
> Happens also in MSFS 2020 in DX11 mode. As the VRAM gets high, stutters and then freeze. Sunshine actually recovers after some 30s but for that it does a restart i guess (in service mode) but when I log back in it crashes again. Only soution is to reduce VRAM usage by changing settings in game.

Probably fixed in latest nightly https://github.com/LizardByte/Sunshine/releases/tag/nightly-dev

**Comment by fsufitch** [comment]
👍 +1, affected here too.

Contributing to the investigation: VRAM fullness causing instability also affects other programs. In my case, streaming Jedi: Survivor through Discord also exhibits some of the same problems (video freezes, audio continues). Discord's entire AV system goes on to crash and only recovers by restarting it.

**Comment by fsufitch** [comment]
Just tested with nightly, and the problem still happened after about 10 minutes of play in Jedi: Survivor. 
![image](https://user-images.githubusercontent.com/4542769/235540396-18811898-a654-4ad4-b2f5-417450d59d12.png)

**Comment by cgutman** [comment]
Try this build (need to be signed into your GitHub account to access) from #3674: https://github.com/LizardByte/Sunshine/suites/12712442439/artifacts/682537209

**Comment by Nonary** [comment]
> Try this build (need to be signed into your GitHub account to access) from #1150: https://github.com/LizardByte/Sunshine/suites/12712442439/artifacts/682537209

Sadly, confirmed it is still freezing.

You can use my test script https://github.com/LizardByte/Sunshine/issues/1031#issuecomment-1504445256

All it needs is Visual Studio C++ build tools and CUDA API
https://developer.nvidia.com/cuda-downloads?target_os=Windows&target_arch=x86_64

And then numpy and pycuda

(You might have to reboot after installing CUDA API)

**Comment by ns6089** [comment]
Even with https://github.com/LizardByte/Sunshine/pull/1150 textures are still potentially evictable, that's just how Direct3D 11 is designed.
> A resource that is set to the maximum priority, DXGI_RESOURCE_PRIORITY_MAXIMUM, is only evicted if there is no other way of resolving the incoming memory request. The Windows Display Driver Model (WDDM) tries to split an incoming memory request to its minimum size and evict lower-priority resources before evicting a resource with maximum priority.

So yeah, synthetic tests can still easily trigger it, but the situation with freezes during game streaming may have improved.

**Comment by neon12345** [comment]
Reading this: [porting-from-direct3d-11-to-direct3d-12](https://learn.microsoft.com/en-us/windows/win32/direct3d12/porting-from-direct3d-11-to-direct3d-12) it seems ffmpeg uses "committed resources" with e.g. ID3D11Device_CreateTexture2D. But there are also "reserved resources" in DX11 called "tiled resources" that are managed manually and probably will not be evicted.

**Comment by cgutman** [comment]
It's unclear to me that eviction is even the cause for this (at least eviction of any resource we control). With #1150, it should only be evictable if all other lower priority resources have been evicted and I highly doubt we've got 16 GB of VRAM full of `DXGI_RESOURCE_PRIORITY_MAXIMUM` resources.

Even if our resources are evicted, the WDDM memory manager should page them back into VRAM transparently when we need them next (evicting another equal or lower priority resource to make room, if needed). Heavy paging activity should only result in poor performance or a memory allocation error at worst, not an application hang.

**Comment by ns6089** [comment]
> With https://github.com/LizardByte/Sunshine/pull/1150, it should only be evictable if all other lower priority resources have been evicted and I highly doubt we've got 16 GB of VRAM full of DXGI_RESOURCE_PRIORITY_MAXIMUM resources.

We've got only one freeze report for that pull request, and there it can be actually happening depending on how nvidia implemented their cuda allocations (quick googling got me nothing).
Also from the original comment with the synthetic script that was used (24GB is total gpu memory):

> Using that at desktop and browsing the internet, the moment my VRAM exceeded 24GB it froze immediately and did not recover even after the VRAM was released.

No idea why if doesn't recover, guess I can try investigating it further after I'm done with the other things.
From the top of my head:
- Reproduce with the cuda script
- Check for [EvictAllocation events](https://developer.nvidia.com/content/are-you-running-out-video-memory-detecting-video-memory-overcommitment-using-gpuview)
- Try directly manipulating last given duplication texture with saved pointer in case duplication api starts returning timeouts

**Comment by Nonary** [comment]
I installed the latest Windows Updates this month and noticed that Sunshine is no longer freezing when exceeding VRAM. It could be that this was a DXGI capture bug all along (since some users reported same issue with Discord and they use DXGI to capture).

I will do more testing, for now if you have Windows 11 22H2 and this issue, please install latest update and confirm if you've noticed freezes have gone away as well.

**Comment by profucius** [comment]
I would like to +1 that this issue is happening to me currently with Diablo 4. I've tried multiple things to ease the pain, but the only solution so far to recover is to force exit in Moonlight and reconnect. But by that time, my character is already dead.

I'm happy to help runs tests and provide debug logs from my end. I have a 3080 original 10gb, i9-9900k, 32gb ram. Anything I can do let me know.

**Comment by ylor** [comment]
Another +1. Diablo IV has been the worst offender, but also happens with other titles. For the time being I've been using Steam Link to stream to iOS/macOS/tvOS but image quality is much worse there.

I also noticed that the stream still freezes even if I use non-GPU encoders like AMF (CPU encoding). Not sure, if that's a helpful data point.

Specs:
- CPU: 7800X3D
- GPU: 3080 (10GB 😢)
- RAM: 64GB DDR5

**Comment by Spiricore** [comment]
Chiming in with the same issues.

i7-12700K
32GB DDR4
RTX 3080

**Comment by Michoko92** [comment]
Having the same issue here, especially with Jedi Survivor. I can monitor VRAM usage, and when it reaches the limits of my VRAM (12 GB), the stream will freeze. I can still hear the audio for a little while, then the PC will crash and reboot.

i7-12700k
32 GB DDR4
RTX 4070
HAGS off
Windows 10
Nvidia Driver 531.79

**Comment by ns6089** [comment]
Recently OBS tried to fix similar issue by decreasing gpu scheduling priority of the process (it appears there's a bug in NVIDIA drivers)

Here's sunshine test build with similar change applied https://github.com/LizardByte/Sunshine/suites/14061415283/artifacts/785859422

Feedback will be appreciated, since I can't easily reproduce the freeze on my system.

**Comment by Michoko92** [comment]
> Recently OBS tried to fix similar issue by decreasing gpu scheduling priority of the process (it appears there's a bug in NVIDIA drivers)
> 
> Here's sunshine test build with similar change applied https://github.com/LizardByte/Sunshine/suites/14061415283/artifacts/785859422
> 
> Feedback will be appreciated, since I can't easily reproduce the freeze on my system.

Wow, sir, thank you so much for your work, I think your change is helping a lot! It could even fix a broader bug in the NVidia drivers, as it seems to fix crashes even when VRAM is not completely saturated (more info below).

First, my specs:
i7-12700k
32 GB DDR4
RTX 4070 12 GB VRAM
HAGS on
Windows 10
Nvidia Driver 531.79

First test: Jedi Survivor
Put all settings to Epic, RT On, no FSR, to maximize VRAM usage. I reached max 11.3 GB of VRAM consumption, but generally it made my game crash at some point while exploring the open area of Koboh. Here, it seems your build doesn't crash, despite trying to load as much textures as possibles in different areas.

Second test: Resident Evil 4
Yesterday, I had a systematic crash at beginning of chapter 4. My VRAM was definitely not saturated (maybe only 7-8 GB were used), but the system kept crashing in this area. Had to disable RT and lower some settings to bypass the crash.
So I just tested with your build, pushing settings even a bit higher than yesterday, and no crash in the same area!

That's why I said above that it might actually fix a broader issue, as I had crashes with those DX12 games even when VRAM was not fully saturated (but lowering VRAM usage was generally reducing / fixing the crashes). 

I'll do further tests, but if I don't report more crashes here, it will mean your implementation definitely improves / solves things in my case. A HUGE thank you for your amazing work and support! Cheers!

**Comment by ylor** [comment]
@ns6089 After installing your test build using latest Nvidia drivers. I have also been unable to induce a freeze using Diablo IV at Ultra settings streaming 120 FPS in 4K HDR.

My VRAM budget is 10GB and Diablo IV ultra textures notoriously exceed that value.

Thank you so much for your efforts ❤️

**Comment by Stono** [comment]
So I went to try this today, full of hope! But i have the strangest bug with this prerelease, my xbox controller (via bluetooth) isn't working properly.  if i press left and right on the left thumbstick it goes up and down!   Can't sus it out!

**Comment by Lucide** [comment]
See https://github.com/LizardByte/Sunshine/issues/1407#issuecomment-1624793437

**Comment by Stono** [comment]
Ahhh @Lucide thanks, does that have this fix in it though?

**Comment by Lucide** [comment]
Right, it shouldn't, oops
Also, the linked artifacts might have expired.
The linked pull request mentions xbox 360 controller emulation, try to select DS4 controller on this build,
maybe it's a workaround.

**Comment by Stono** [comment]
@Lucide that fixes the up/down/left/right but then all my other key mappings are messed up in game haha.  Guess i'll just have to wait for these things to merge into the mainline.

**Comment by profucius** [comment]
> Right, it shouldn't, oops Also, the linked artifacts might have expired. The linked [pull request](https://github.com/LizardByte/Sunshine/pull/1426) mentions xbox 360 controller emulation, try to select DS4 controller on this build, maybe works as a workaround.

If I understand this correctly, the build mentioned by ns6089 above has the VRAM fix, but not the controller fix. And the pull request build you mentioned has the controller fix but not the VRAM fix?

Because I tried the first one and the controller wasn't working correctly (left/right was up/down), and then I updated to the newest nightly yesterday, and the controller was fixed but the VRAM issue happened again while playing Diablo.

If I am understanding this correctly, then should we just be patient and wait for the fixes to be merged in a future nightly?

**Comment by Lucide** [comment]
I see the controller-related pr got merged into nightly yesterday.
In the pr related to this issue, ns6089 [said](https://github.com/LizardByte/Sunshine/pull/1197#issuecomment-1627355666)

> Can somebody hit the CI again please? The nightly version has bugs that causing problems for testers, so I've rebased it again stable.

In sunshine stable xbox controller emulation works, so you should soon be able to try the freeze fix. He requested a new automated build.

**Comment by ns6089** [comment]
Freeze fix on top of latest stable version https://github.com/LizardByte/Sunshine/suites/14153789829/artifacts/793903266

**Comment by profucius** [comment]
> Freeze fix on top of latest stable version https://github.com/LizardByte/Sunshine/suites/14153789829/artifacts/793903266

This version still has the issue with the controller inputs being mis-mapped. Nvidia Shield controller via Moonlight, pressing left/right directions moves the cursor up/down. Pressing up/down moves the cursor nowhere. It's the same issue as mentioned earlier as a response to a fix for the VRAM fix implementation.

**Comment by cgutman** [comment]
@profucius Your issue sounds different than the Xbox 360 emulation regression in nightly. Does it happen with Sunshine v0.20? You are using an actual Nvidia Shield controller? Are you using it on PC or Android?

**Comment by profucius** [comment]
> @profucius Your issue sounds different than the Xbox 360 emulation regression in nightly. Does it happen with Sunshine v0.20? You are using an actual Nvidia Shield controller? Are you using it on PC or Android?

v20 release as of May 2023 does not have this controller issue for me, but using these nightly builds does.

Official Nvidia Shield running Android. Official Nvidia Shield Controller. Sunshine server installed on Windows 11. Again never had any of these issues (except for the Vram inflicted crash issue playing Diablo) until I started installing these nightly builds.

**Comment by cgutman** [comment]
Hmm, I just tested with my Shield and Shield controller and I'm not seeing those issues here with Sunshine nightly on Win11. I tried both DS4 and Xbox 360 emulation. D-Pad and sticks are moving in the correct directions on the emulated gamepad.

Please file a bug for this so we can collect logs and have further discussion there.

**Comment by Dijkie85** [comment]
Is the fix already live in the nightly release? I'm on windows 7/18 nightly and still running into this issue :(

**Comment by ns6089** [comment]
> Is the fix already live in the nightly release? I'm on windows 7/18 nightly and still running into this issue :(

It should be in the nightly, yes. And I would call it more of a workaround than a fix, since the problem seems to be in nvidia drivers and we can do nothing about them.

Please update to the latest drivers and if the issue continues to happen provide us with more information. We're interested in GPU model, streaming resolution-fps-bitrate, OS version, whether [Hardware Accelerated GPU Scheduling](https://devblogs.microsoft.com/directx/hardware-accelerated-gpu-scheduling/) is enabled, and the exact situation (game) which leads to a freeze.

**Comment by Dijkie85** [comment]
Thanks, will test a bit more and report back. For what it's worth, using GFE works just fine (but with less fps: 45@1080 vs 60 w/ sunshine, and the gamepad is less snappy). Forgive my ignorance but isn't it possible to emulate in Sunshine the way GFE handles VRAM? Because I can't help to think that the driver is the same in both scenarios.
 I'm on the latest 536.67 (7/18/23), on a gtx 1650 ti (Asus tuf notebook) playing Astlibra, which is just a 2d, not hardware intensive game. HAGS turned on as I read that didn't cause more trouble.

**Comment by ns6089** [comment]
> Forgive my ignorance but isn't it possible to emulate in Sunshine the way GFE handles VRAM? Because I can't help to think that the driver is the same in both scenarios.

GFE uses completely different capture API that they since deprecated years ago. Consequently they're using full CUDA processing path while we're stuck with Direct3D 11.
I wish it was as simple as emulating what they're doing 😞

**Comment by Dijkie85** [comment]
Ah I get it! Thanks for all the hard work!

**Comment by ReenigneArcher** [comment]
>  I'm on windows 7/18

Please update your OS. Windows 7 has been unsupported for 3.5 years and Windows 8.1 has been unsupported for 6 months.

**Comment by Dijkie85** [comment]
I'm sorry I meant to say I'm running the windows installer nightly from 7/18. I'm on windows 11 with the latest updates

**Comment by ns6089** [comment]
In the current nightly lowering HAGS scheduling priority is made into config option on NVENC encoder config page, with realtime (potentially freezing) being the default. Using "high" by default was not feasible, since it affected capture frame rate when gpu was 100% loaded.

**Comment by noahtallen** [comment]
I was running into this issue in Jedi Survivor, so I installed the current nightly (`0.20.0.b00d7af36929f70cdb887dc0adba98b1110a503e`) to try to fix it. Unfortunately, I think this is still a problem. (I couldn't find any documentation on how to install the latest nightly, so I installed the artifact from the latest commit in the nightly branch.)

I'm streaming from Sunshine to Moonlight on the Steam Deck, which plays the Sunshine stream on my TV. I currently have the stream set to 4k@60fps (150Mbps bitrate) in Moonlight, partly because my TV and network support it, and partly to test the limits of streaming to my TV.

I'm still very early game, and the freeze occurred pretty quickly in the tutorial, when it tells you how to do a mid-air double jump. After installing the nightly, the freeze occurred right after meeting your initial companion, where there is a rope you can cut to open a shortcut. I also experienced the freeze when opening the mapa

At this point, I tried lowering some settings (running 1440p in-game with DLSS disabled, and a 1440p 100Mbps stream from Moonlight). I seemed to get somewhat further in the game, but I ultimately still experienced the freeze at least two more times in the initial mission before I gave up playing through Sunshine.

The biggest issue IMO is that Sunshine cannot recover from this issue until you quit out of the game, meaning you lose progress in the game unless you decide to continue to the next checkpoint back on your PC.

In terms of how to reproduce, play a new game of Jedi Survivor through starting tutorial mission for like 30min on high settings and it'll probably happen.

#### Hardware info:
- RTX 3080 with 10GB dedicated VRAM.
- RX 5800x3d
- 32GB RAM

#### Software versions:
- Latest NVIDIA drivers (`537.13`)
- Windows 11 (version `22H2`, build `22621.2215`)
- Sunshine nightly (`0.20.0.b00d7af36929f70cdb887dc0adba98b1110a503e`)
- HAGS enabled

#### Game settings:
- Jedi Survivor (patch 7)
- 4k, mostly epic settings (testing the limits)
- DLSS set to quality

**Comment by cgutman** [comment]
Did you adjust the new HAGS option mentioned in the previous comment? The realtime preset (which triggers the driver bug) is the default and you can lower it to high if your system is hitting the HAGS hang. You can also turn off HAGS in Windows too in order to avoid the driver hang.

> The biggest issue IMO is that Sunshine cannot recover from this issue until you quit out of the game, meaning you lose progress in the game unless you decide to continue to the next checkpoint back on your PC.

How are you launching the game? If you're streaming the Desktop and launching the game, Sunshine should be able to recover by just disconnecting the stream and waiting a few seconds for it to detect the hang and restart itself.

**Comment by noahtallen** [comment]
I just tested the HAGS option in the previous comment, and it definitely made the stream unplayable as noted in the PR. Very choppy, even after downgrading the stream quality to 1440p 40mbps. 

I've been launching the game via Steam Big Picture, which is one of the default options in Moonlight. But I can try what happens when I start it from the desktop.

**Comment by noahtallen** [comment]
I'm also not sure about disabling HAGS -- I've heard it's necessary for DLSS, which I'd like to keep using (especially in this poorly optimized game 😅 )

**Comment by cgutman** [comment]
Can you describe exactly what happens when you say you need to quit the game to resolve the hang? Since you're not directly launching Jedi Survivor from Sunshine, it should stay running even if Sunshine is restarted. If you disconnect the session and wait about 10 seconds, you should see the play icon disappear from the Steam tile in Moonlight and then you can use the Desktop app to stream again without starting any new apps.

> I'm also not sure about disabling HAGS -- I've heard it's necessary for DLSS

HAGS is only necessary for DLSS 3's Frame Generation, not the regular AI upscaling. If you're not using Frame Generation, HAGS is not required.

**Comment by noahtallen** [comment]
Good info -- sounds like disabling HAGS would be an option then.

Still testing this out to see if I can dig up any helpful info:

So I tried launching via desktop mode. I had the game open already on my PC, so I simply launched 'desktop' from Moonlight. After playing smoothly for about 15min, I experienced another freeze. I waited about 30s, but the stream did not recover. I then disconnected from the stream in Moonlight. However, when I do this, I cannot reconnect. The PC shows up as unavailable (with the "wake PC" option in Moonlight), and I get time outs trying to reconnect to the desktop.

Essentially, it seems like Sunshine doesn't recover. And when I go back to my PC, I try to manually restart it, but this doesn't solve the issue. (In fact, the local website is unavailable, indicating Sunshine has issues starting up.) 

However, after I quit the game from my PC, Sunshine will start up again and I can reconnect in Moonlight. But it doesn't recover while the game is still open, and I going back to my PC to restart Sunshine manually doesn't work either.

The freeze is also weird -- when it happens, audio still works, as does the controller. I can actually watch my PC's monitor respond to my controls -- it's just that the video stream in Moonlight is completely frozen. When I disconnect in Moonlight as described above, I cannot reconnect though. It's odd that the connection still partially works, but the PC is considered "offline" after disconnecting in Moonlight.

Also, twice when this has happened, the game itself crashed maybe 30s after the freeze in Sunshine! Most of the freezes, this did not happen, but the error is interesting. The important part is `DXGI_ERROR_DEVICE_REMOVED with Reason: DXGI_ERROR_DEVICE_HUNG`.

![image](https://github.com/LizardByte/Sunshine/assets/6265975/552ede0e-321c-4bb9-bf37-d13d4d67ca48)

> If you disconnect the session and wait about 10 seconds, you should see the play icon disappear from the Steam tile in Moonlight and then you can use the Desktop app to stream again without starting any new apps.

I did wait longer than 10s to attempt a reconnect, but it's possible I didn't wait long enough. That said, I think Sunshine is struggling to restart while the game is still open.

**Comment by noahtallen** [comment]
Disabling HAGS has definitely fixed it. I did experience one freeze after a game crash, but I was able to recover quickly.

**Comment by noahtallen** [comment]
I don't know if it'd be fair to say that either this issue or #863 are actually fixed unless new commits have been added. If you need HAGS (totally possible with newer GPUs and games), it's still an issue. The fix isn't usable (as noted above) due to capture framerate

**Comment by ReenigneArcher** [comment]
@noahtallen multiple PRs have been merged that fix these issues.

We label issues as "fixed" when it has been fixed in our development branch (aka "nightly") and close the issues when we make a new stable release.

**Comment by noahtallen** [comment]
Right -- however, the report I made above (https://github.com/LizardByte/Sunshine/issues/1031#issuecomment-1712647163) was running on a nightly build which should have included all the builds. As far as I can tell, disabling HAGS is the only real way to fix the problem (and that might not be an option for many). I do appreciate and understand these are tricky to fix, but I'm just saying this might still need to be tracked and not closed :) 

Or, maybe this issue related to VRAM is fixed, but #863 is not since the workaround in settings has the side effect of making the stream unusable.

**Comment by Nonary** [comment]
> Right -- however, the report I made above ([#1031 (comment)](https://github.com/LizardByte/Sunshine/issues/1031#issuecomment-1712647163)) was running on a nightly build which should have included all the builds. As far as I can tell, disabling HAGS is the only real way to fix the problem (and that might not be an option for many). I do appreciate and understand these are tricky to fix, but I'm just saying this might still need to be tracked and not closed :)
> 
> Or, maybe this issue related to VRAM is fixed, but #863 is not since the workaround in settings has the side effect of making the stream unusable.

Truthfully I only opened this issue because I wanted to make sure it would show up in searches in Google and GitHub. At the time I didn’t exactly know the issue is unsolvable on our end due to it being an issue with the NVIDIA drivers themselves.

As long as people understand this issue is out of our control and is realistically unsolvable I am not opposed to keeping this issue open. @ReenigneArcher your thoughts on that? Maybe we add a new label to make it obvious it’s an external issue?

**Comment by ReenigneArcher** [comment]
I think we'll probably close it for the following reasons.

- The situation has been improved by some code improvements on our side as well as fixes outside our control. Maybe it's not 100% solved, but some of the games reported in your original comment have been reported by some users to be perfectly playable now.
- The issue is kind of un-actionable on our side.
- I don't want Sunshine to turn into a repo with hundreds or thousands of open issues as it becomes too overwhelming to manage. The ~70 we have open now is already overwhelming.
- Contributors are constantly looking for ways to improve performance, memory issues, etc.

We can leave it open if there is a strong objection to closing it, but I don't see much value in it unless a dev wants to investigate further how we can resolve it better on our end.

@noahtallen P.S. Nightly download link is available on our new Sunshine specific project site: https://app.lizardbyte.dev/Sunshine

Downloading from any other source is not recommended unless a dev directs you there.

**Comment by Nonary** [comment]
The new changes make it an option to change priority, it's not on by default because a lot of users reported issues. These were mostly users who were pushing the limits of their card though. I am only suggesting to keep this open as this will be a common issue for users (noticing that their stream is freezing). 

When I created this issue, the amount of "why is my stream freezing" in the discord dropped dramatically. I think there is rather large crowd who finds out about this issue solely on searches and checking the open issues.

As for the other open issues I will work on closing those out if they're not actually an issue...

**Comment by noahtallen** [comment]
That's a fair point @ReenigneArcher. I think it'd be good to document this on the site too (particularly that disabling HAGS might be required depending on the game.)

> These were mostly users who were pushing the limits of their card though.

the issue is that many modern games will just do this without any intervention. I'm not trying to be a power user, but the popular AAA game I played happens to really eat up resources. Which that anyone playing a certain set of games has a decent chance of running into this issue and wants to quickly find the workaround

**Comment by hot22shot** [comment]
> That's a fair point @ReenigneArcher. I think it'd be good to document this on the site too (particularly that disabling HAGS might be required depending on the game.)

I second that, FYI I had this issue with my RTX 3090 TI (yes, with 24GB of VRAM) and a heavily modded Cyberpunk 2077 (>800 mods) in 4k.
And it bothered my for a while as I couldn't figured out what was causing this.
So if this could be documented somewhere it would be fine.

**Comment by ReenigneArcher** [comment]
Anyone is free to open a PR to adjust/improve documentation.

**Comment by Travillion** [comment]
So I just want to clarify, as I believe I am consistently getting this in Jedi: Survivor. Is best practice to disable HAGS, or lower visual settings, or both? I saw several comments in the HAGS bug report indicating disabling HAGS was no longer required because the issue was fixed. It seems from reading this thread the issue is not really fixed, just fixed as much as the sunshine team can do, which should be an improvement.

**Comment by ryanteck** [comment]
I've experienced this issue a few times, first time I've experienced it but only been in Jedi Survivor. I tried decreasing quality settings first without luck, disabiling HAGS seems to have resolved it for now.

Even though I was using desktop to stream sometimes it would not recover, partly because the game then was crashing simultaneously. However to re-iterate it seems to be just Jedi Survivor so maybe just poorly optimised game?

**Comment by Lucide** [comment]
No.
I get freezing on my hags-enabled system in other demanding titles like the two sony spiderman. The system has an rtx 2080 which is a 8GB card.

**Comment by wirytiox** [comment]
> At this point I'm starting to think all these new dx12 games with unevictable vram allocations are capable of breaking windows side of desktop duplication api, so it starts returning timeouts instead of surfaces. Needs testing, and if confirmed we can simply reinit capture on prolonged timeouts.

the freezes happened to me while playing skyrim and when using comfyui with PDXL (basically a lot of ram ussage.)

**Comment by JaleSwiftpaw** [comment]
Confirming this issue as well in games with very high VRAM usage such as Final Fantasy VII Rebirth on a RTX 2060 Super 8 GB. The only way Sunshine would remain streaming is by setting the texture quality to low so it uses as much as 6~7 GB of VRAM.

**Comment by Tikilou** [comment]
Same for me with FF7 Rebirth on GTX 1660TI, it's fine on Linux, but not on Windows.

**Comment by HewellP** [comment]
Confirm still freeze with HAGS disabled with a RTX 3060 with 40% VRAM load.

**Comment by Nonary** [comment]
This issue will be fixed once #4066 is finished.

**Comment by L1ttleFlyyy** [comment]
@Nonary Hi, if you don't mind, could you brief explain why the new capture API will solve this freezing issue?

**Comment by Nonary** [comment]
> [@Nonary](https://github.com/Nonary) Hi, if you don't mind, could you brief explain why the new capture API will solve this freezing issue?

The issue occurs primarily on NVIDIA drivers when HAGS (Hardware-Accelerated GPU Scheduling) is enabled. With DXGI desktop duplication, every captured frame is parked in VRAM until you release it. If VRAM is full (for example, because the game you’re streaming has consumed most of it), the driver can’t allocate another capture texture. As a result, AcquireNextFrame() may block indefinitely, leading to a deadlock.

Windows Graphics Capture, on the other hand, operates at a higher level by capturing frames through the Desktop Window Manager (DWM). When VRAM is exhausted, WGC can fall back to using shared system memory (RAM) for its staging buffers. This allows capture to continue... albeit potentially with dropped frames or reduced performance, without deadlocking, because it isn’t as tightly coupled to the NVIDIA driver or strictly reliant on VRAM for every frame.

**Comment by MemeTao** [comment]
> > [@Nonary](https://github.com/Nonary) Hi, if you don't mind, could you brief explain why the new capture API will solve this freezing issue?
> 
> The issue occurs primarily on NVIDIA drivers when HAGS (Hardware-Accelerated GPU Scheduling) is enabled. With DXGI desktop duplication, every captured frame is parked in VRAM until you release it. If VRAM is full (for example, because the game you’re streaming has consumed most of it), the driver can’t allocate another capture texture. As a result, AcquireNextFrame() may block indefinitely, leading to a deadlock.
> 
> Windows Graphics Capture, on the other hand, operates at a higher level by capturing frames through the Desktop Window Manager (DWM). When VRAM is exhausted, WGC can fall back to using shared system memory (RAM) for its staging buffers. This allows capture to continue... albeit potentially with dropped frames or reduced performance, without deadlocking, because it isn’t as tightly coupled to the NVIDIA driver or strictly reliant on VRAM for every frame.

I wonder why Nvidia still hasn‘t fixed this bug after two years.

**Comment by profucius** [comment]
> I wonder why Nvidia still hasn‘t fixed this bug after two years.

Nvidia doesn't care about its graphics division anymore, only AI tech which is the lion's share of its profits. Won't be surprised if we don't even get a next-gen cards announcement this year or next. 

I don't mean to sound defeatist but I've given up on them fixing a lot of the ongoing issues while the AI bubble is still inflated.

**Comment by umutv** [comment]
up to date 28.07.2026 still I have this bug! it's not fixed yet!
